osu.Server.Spectator/Hubs/StatefulUserHub.cs

```diff
@@ -2,94 +2,106 @@
 
         protected StatefulUserHub(IDistributedCache cache)
         {
-            this.Cache = cache;
         }
 
+        protected static KeyValuePair<long, TUserState>[] GetAllStates() => ACTIVE_STATES.GetAllEntities();
+
         /// <summary>
         /// The osu! user id for the currently processing context.
         /// </summary>
         protected int CurrentContextUserId => int.Parse(Context.UserIdentifier);
 
-        public override Task OnConnectedAsync()
+        public override async Task OnConnectedAsync()
         {
             Console.WriteLine($"User {CurrentContextUserId} connected!");
 
-            return base.OnConnectedAsync();
-        }
+            // if a previous connection is still present for the current user, we need to clean it up.
+            await cleanUpState(false);
 
-        /// <summary>
-        /// Called when a user disconnected, providing their last state.
-        /// </summary>
-        /// <param name="exception">A potential error which caused the disconnection.</param>
-        /// <param name="state">The last user state. May be null. This is automatically cleared on disconnection.</param>
-        protected virtual Task OnDisconnectedAsync(Exception exception, TUserState? state) => Task.CompletedTask;
+            await base.OnConnectedAsync();
+        }
 
         public sealed override async Task OnDisconnectedAsync(Exception exception)
         {
             Console.WriteLine($"User {CurrentContextUserId} disconnected!");
 
-            var state = await GetLocalUserState();
-
-            await OnDisconnectedAsync(exception, state);
-
-            // clean up user on disconnection
-            if (state != null) await RemoveLocalUserState();
-
-            await base.OnDisconnectedAsync(exception);
+            await cleanUpState(true);
         }
 
-        protected async Task UpdateLocalUserState(TUserState state)
+        private async Task cleanUpState(bool isDisconnect)
         {
-            ACTIVE_STATES.TryRemove(CurrentContextUserId, out var _);
-            ACTIVE_STATES.TryAdd(CurrentContextUserId, state);
-
-            await Cache.SetStringAsync(GetStateId(CurrentContextUserId), JsonConvert.SerializeObject(state));
+            ItemUsage<TUserState>? usage;
+
+            try
+            {
+                usage = await ACTIVE_STATES.GetForUse(CurrentContextUserId);
+            }
+            catch (KeyNotFoundException)
+            {
+                // no state to clean up.
+                return;
+            }
+
+            using (usage)
+            {
+                if (usage.Item != null)
+                {
+                    bool isOurState = usage.Item.ConnectionId != Context.ConnectionId;
+
+                    if (isDisconnect && !isOurState)
+                        // not our state, owned by a different connection.
+                        return;
+
+                    await CleanUpState(usage.Item);
```

Comment:
As I attempted to (and probably failed to) convey earlier in https://github.com/ppy/osu-server-spectator/pull/25#discussion_r550166581, if this virtual `CleanUpState()` ever throws (as the `MultiplayerHub` one can), I think the `usage.Destroy()` lower down will never be called. Should probably be

```csharp
try
{
	if (usage.Item != null)
	// ...
}
finally
{
	usage.Destroy();
	usage.Dispose();
}
```

Reply:
You bring up another interesting point: should `usage.Dispose` call destroy implicitly if it's not already destroyed? or are we happy with the system being closed enough that this is not a requirement?

Reply:
Wait, dispose calling destroy? Dispose just frees the lock - otherwise all entities would get immediately deleted after usage. In fact `Destroy()` includes `Dispose()` now. In fact in the above I missed the fact that the `Dispose()` is gratuitous above - it's just a no-op at that point.

Reply:
Back to front, oops.